README.md
[![CircleCI](https://circleci.com/gh/calgacus/TlkEdit-EE.svg?style=svg)](https://circleci.com/gh/calgacus/TlkEdit-EE)

# TlkEdit-EE

This is a continuation of the TlkEdit project found at https://neverwintervault.org/project/nwn2/other/tool/2datlkgff-editor-aka-tlkedit2

Now on version 14.0.6

Any input is welcome, I hope to be able to expand on this tool.  I do not plan to support the launch4j part that creates the *.exe 32bit-java files.

My main foocus is supporting this project for NWN:EE but hopefully I will not have to break any support for the other games (NWN 1.69, NWN2, Witcher).

If you want spell-check for other languages there are some extra "myspell" format dictionary files at -> http://download.services.openoffice.org/contrib/dictionaries/
Download the desired ones to the /dict folder which should be in the same folder in which the tlkedit.jar file is present,
then update ./dict/dictionaries.properties to include the needed code like the others in the same file.


Debug
-----
in vscode the main file is  \TlkEdit-EE\src\main\java\org\jl\nwn\editor\EditorServer.java

Select the menu Run->Start Debugging ( or hit F5)


Run
-----
Install java, and if you want to compile from source maven and, if on windows, a unix style shell, eg git-bash.
In project root run ./tlkedit.sh or ./tlkedit2.sh for the NWN2 version.

Build
-----
Install java, maven and, if on windows, git-bash

First, install dependencies into local maven repository. This is temporal solution until
dependencies will be upgraded to one that exists in one of public maven repositories:

```bash
$ mvn install:install-file -Dfile=lib/jmyspell-1.0.0-beta1.jar -DgroupId=tlkedit -DartifactId=jmyspell -Dversion=1.0.0-beta1 -Dpackaging=jar -DlocalRepositoryPath=./lib
```

Then build project as usual:

```bash
mvn install
```

### Troubleshooting
If you got error:
```bash
$ mvn install
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building TlkEdit-EE 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[WARNING] The POM for tlkedit:jmyspell:jar:1.0.0-beta1 is missing, no dependency information available
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.392s
[INFO] Finished at: Mon Sep 30 19:08:08 YEKT 2019
[INFO] Final Memory: 4M/123M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project TlkEdit-EE: Could not resolve dependencies for project tlkedit:TlkEdit-EE:jar:1.0-SNAPSHOT: The following artifacts could not be resolved: tlkedit:jmyspell:jar:1.0.0-beta1: Failure to find tlkedit:jmyspell:jar:1.0.0-beta1 in file://D:\Projects\NWN\TlkEdit-EE/lib was cached in the local repository, resolution will not be reattempted until the update interval of local-jars has elapsed or updates are forced -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
$
```
it means, that you already tried to build project without success and maven cache
unsuccesfull state of dependency resolution. Just add switch `-U` to maven invocation
to force maven re-check dependencies:

```bash
$ mvn -U install
```

Credits
=======
TlkEdit wouldn't be the same without
* The [SwingX](http://swingx.dev.java.net/) library
* [JMySpell](http://jmyspell.javahispano.net/)

Other peoples' work included in this release:
* TlkEdit.exe is a [launch4j](http://launch4j.sourceforge.net/) native java launcher
* The icons are part of the 'Nuvola' icon set, created by David Vignoni for KDE. See http://www.icon-king.com/.
* Patch version of dlpcpakv01 by Garad Moonbeam (dragonlance races).

NWN1 vs NWN2
============
The tlk and gff format for NWN1 and NWN2 is the same except for the string encoding used.
That means that TlkEdit cannot distinguish between files for NWN1 or NWN2, but opening
a NWN2 tlk file as NWN tlk file will display lots of garbled text. When launched with
TlkEdit2.exe TlkEdit will use NWN2 mode by default. To open or save a file in NWN1 format
you need to select the **All Files - NWN 1** filter in the Open File / Save As dialog.
The mode in which a file will be written is also displayed in the title bar.
The default mode can be set with the `tlkedit.defaultNwnVersion` system property,
i.e. add `-Dtlkedit.defaultNwnVersion=NWN1` or `-Dtlkedit.defaultNwnVersion=NWN2`
to the tlkedit startup file (either `TlkEdit.l4j.ini/TlkEdit2.l4j.ini` or `tlkedit.sh`).

Running TlkEdit
===============
Windows
-------
TlkEdit.exe and TlkEdit2.exe are provided as .EXE launchers for TlkEdit, startup properties
can be found in `TlkEdit.l4j.ini`/`TlkEdit2.l4j.ini` (plain text file).

Linux / OS-X
-------------
See `tlkedit.sh`.

Server Mode
-----------
By default, TlkEdit will try to connect an already running instance so that there can only be
one open TlkEdit instance containing all opened files. This uses TCP/IP so your firewall
software might complain.

Memory Usage
------------
The maximum amount of memory java is allowed to use by default is 64MB. This can be overridden
using the `-Xmx` switch. Both `TlkEdit.lap` and `tlkedit.sh` use this to set the maximum amount
to 128 MB. Memory usage (heap space) is displayed in the TlkEdit status bar, so you can tell
when you are running out of space (some spell checking dictionaries are rather greedy - `de_DE_comb`
needs about 45 MB).

Spell Checking
--------------
Spell checking is provided by [JMySpell], a Java implementation of the OpenOffice spell checker MySpell.
Dictionaries can be downloaded from the [OpenOffice web site][oo_dict].
The site offers dictionaries in both myspell and (newer) hunspell format. JMySpell will only work
with myspell dictionaries. The dictionary zip file should be placed in the `dict` subdirectory
of the TlkEdit dir. Afterwards you need to edit the file `dict/dictionaries.properties` and set
the correct dictionary file for your language code. If you want to use `en_US.zip`
([download][en_US]) or `de_DE_comb.zip` ([download][ed_DE]), the properties are already set correctly.

[oo_dict]: http://wiki.services.openoffice.org/wiki/Dictionaries
[en_US]: http://ftp.services.openoffice.org/pub/OpenOffice.org/contrib/dictionaries/en_US.zip
[de_DE]: http://ftp.services.openoffice.org/pub/OpenOffice.org/contrib/dictionaries/de_DE_comb.zip

Editing
=======
Key bindings
------------
Key bindings (hot keys) can be altered by editing the file `keybinding.properties`
in the `settings` directory. TlkEdit needs to be restarted for the changes to take effect.
On OS-X TlkEdit should map 'control' to the 'command' key, but I couldn't test it.

Editing tlk files
-----------------
Cell editing is stopped by clicking somewhere outside the cell with the mouse or preferably by
pressing ALT+ENTER (you can input line breaks by pressing enter).

Unless you are editing a cell, cut/copy/paste work in entire table rows. This uses an internal
clipboard, so that you can copy tlk entries to other tlk files opened in the editor, but not to
other applications.

### Search & Replace
Pressing Enter or clicking the "find" button will hide the search dialog, unless the "keep dialog"
checkbox is selected. When "keep dialog" is checked you can step through the occurrences of
the search term with "find".

### The "flags" field
This field defines which of the 3 values (String, Sound ResRef and Sound Length) are actually used by NWN.
If you enter a string but set the flags to 0, the string will NOT be used in game. The flags field is set
automatically when you enter something in the "String", "Sound ResRef" or "Sound Length" field, so you can
safely ignore it.

### diff Menu
The tlk editor keeps track of changes to the tlk table. Modified lines are marked with `*`. This mark can be
set/removed manually with the "+*" / "-*" buttons (hotkeys Alt-m / Alt-u)

Menu items
- **"save diff"** : save all modified entries as diff file
- **"merge diff"** : load a diff file. Entries in the diff file will replace existing entries and will be marked as modified
- **"merge dtu"** : load a file created by DialogTLK
- **"discard diff info"** : set all lines to unmodified (this will NOT restore the original tlk entries)
- **"diff overview"** : display a list with the numbers of all modified entries. Selecting a number in the list will display the entry in the editor

(note : there is no default extension for these diff files, i use .tlu)

### tlk lookup
Whenever you edit 2da or Gff files and select a cell with an integer value or a `CExoLocString` node
TlkEdit will try to resolve that value as a StrRef and display the TLK String in the text area in the
TLK lookup dialog. Of course you need to have at least one tlk file opened for this.
If you have more than one tlk file opened in the editor you can select which one is used for tlk lookup
by activating the tab and selecting the appropriate item from the **'tlk lookup'** menu.
By default the first tlk file that's loaded will be used as default tlk table and the second will be used
as user tlk table (i.e. for StrRefs `>16777216`).

Editing 2da files
-----------------
Nothing much to say here ;) If you input strings that contain spaces the editor will try to add
quotation marks as needed. The cut/copy/paste functions use the system clipboard, which means
you can copy 2da lines into another application (like Excel) and vice versa.
Paste works only if the number of columns matches.

When entering StrRefs for custom content you might want to try the *"Edit->Toggle User Flag on StrRef"* menu item.
The **'alter table'** menu item pops up a dialog that will allow you to add, remove (drop) or rename a table column.

Editing gff files
-----------------
Cut/Copy/Paste works on entire subtrees, this uses an editor internal clipboard, so you can copy nodes to another gff file.
The **'find'** function will not find text inside `void` fields and **'replace'** can only be used on `CResRef`, `CExoStrings` and `CExoLocString` substrings.

By default, pressing the `TAB` key will expand / collapse a selected struct or list.

Editing erf files
-----------------
Should be straightforward enough :) Careful when editing ResRefs though: if you rename a resource
and another resource with the same name already exists it will be replaced!
Selecting **'edit resources'** from the **erf** menu will open selected 2da or gff files for editing.
The files will be extracted as temporary files and a new editor panel is added for each.
Everytime you save one of that editor panels the whole erf file will be rewritten.

Encodings (char sets)
=====================
NWN identifies a TLK file's language by looking at the value of byte number `0x08` in the tlk file
(the byte following the file type + version string `"TLK V3.0"`). This byte also determines
the encoding of the file, i.e. the way all those bytes are converted into readable text.
The language to encoding mappings used by TlkEdit are as follows:

|=Code|=Language          |=Encoding NWN 1|=Encoding NWN 2|
|    0|English            |CP1252         |UTF-8          |
|    1|French             |CP1252         |UTF-8          |
|    2|German             |CP1252         |UTF-8          |
|    3|Italian            |CP1252         |UTF-8          |
|    4|Spanish            |CP1252         |UTF-8          |
|    5|Polish             |CP1250         |UTF-8          |
|    6|Unknown            |Unused (CP1250)|UTF-8          |
|  128|Korean             |MS949          |UTF-8          |
|  129|Traditional Chinese|MS950          |UTF-8          |
|  130|Simplified Chinese |MS936          |UTF-8          |
|  131|Japanese           |MS932          |UTF-8          |

These values are used for tlk files and `CExoLocStrings` in gff files.
The mapping used for languages 0-4 is correct as far as I can tell.
However the encoding used can be modified with the `-Dtlkedit.charsetOverride`
command line switch. Values are of the form:
```
<NWNVersion>:<code>:<encoding>[;<code>:<encoding>]+
```
Launching TlkEdit with
```console
java -Dtlkedit.charsetOverride="NWN1:2:ISO-8859-1" -jar tlkedit.jar
```
would change the encoding used for German (2) NWN1 tlk files to ISO-8859-1.
For NWN2 use "NWN2" as version string. The override parameters can also be added
to the `tlkedit.lap` or `tlkedit2.lap` files which contain the startup configuration
for `TlkEdit.exe` and `TlkEdit2.exe`.

The list of encodings supported by Java2 can be found in the documentation under
"Guide to Features"->"Basic Features, Internationalization"->"Supported Encodings"


CHANGELOG.md
[Release 14.0.6] (2022-08-28)
fixed a bug oin the font color of the file tabs
===========================
[Release 14.0.5] (2022-08-28)
recompiled jar to include updates made by shadguy to spells files

============================
[Release 14.0.4] (2019-10-07)
increased allowable size of CExoStrings to 128*1024

=============================

[Release 14.0.3] (2019-10-07)
increased allowable size of CExoStrings to 2*1024

=============================
[Release 14.0.2] (2019-10-07)
=============================
- added circleci build support (Thanks again Mingun)
- changed version.txt to CHANGELOG.md

[Release 14.0.1] (2019-10-06)
=============================
- Map escape to cancel action for 2da cels (Thanks Mingun {on github})
- Cleaned up and modernize java code (Mingun)
- Changed way tlkedit version number is displayed (now only found in version file)

[Release 14.0.1]: https://github.com/calgacus/TlkEdit-EE/commit/cf63cf2c3a9fa4218e8f20060709bdb99b291f64


[Release 14.0] (2019-10-02)
=========================
- Host project on github at https://github.com/calgacus/TlkEdit-EE
- Allow column resizing for 2DAs
- Fix a bug which prevent some column popups in some 2DAs
- Switch to a maven build (Mingun)
- Cleaned up and modernize java code (Mingun)
- Include build steps in README (Mingun)

[Release 14.0]: https://github.com/calgacus/TlkEdit-EE/commit/17ca29453f6345c72dbb36761c5fc2c6c358a9d7


Release 13c (2007-07-15)
========================
- Updated to a recent swingx build
- Now requires Java 6
- Replaced janel launcher with launch4j

TLK editor
----------
- Added filter function using regular expressions


Release 13b (2007-01-14)
========================
TLK/2DA editor
----------
- Copy & paste StrRefs of selected tlk entries into a 2da table

2DA editor
----------
- Loading of special cell editors & renderers is now based on NWN version


Release 13a (2006-12-13)
========================
- File version ( NWN1 or NWN2 ) can be selected in the open/save dialogs

ERF editor
----------
- Open & extract NWN2 mod & erf files, creating erf files is untested

GFF editor
----------
- All `.ifo` and `.bic` files should work now (error caused by `CExoLocSubstring`s with language id -1,
  added as language `GffToken`)
- Open `.ros` files

TLK editor
----------
- Display StrRefs as hexadecimal values


Release 13 (?)
==============
- (quick&dirty) fixes for NWN2 compatibility
- Support for 2DA V2.b files (KOTOR 1&2)
- Support for Vector type in gff files (KOTOR 1&2)


Release 12c (2006-10-22)
========================
- Fixed some memory issues

GFF editor
----------
- Empty lists were broken (bad code found by EPOlson)


Release 12b
===========
2DA editor
----------
- Fixed 'paste' action (broken in Release 12)


Release 12
==========
- TlkEdit now requires Java 1.5 (a.k.a Java 5)
- Uses `swingx` swing extensions (http://swingx.dev.java.net)
- Undo / Redo in TLK, GFF and 2DA editors
- Spell checking for TLK and GFF files
- Added menu items for most functions
- Editor toolbar is displayed in main window instead of editor panel
- Key bindings can be altered by editing the file `settings/keybindings.properties`
- Included TlkEdit.exe launcher for windows
- Drag&Drop : open files by dragging them onto the editor toolbar
- File browser tree has been removed
- TLK lookup is now an extra dialog window

GFF editor
----------
- Editor control is now a tree table which should make editing simpler
- Bugfix: data type `CHAR` was treated as unsigned value,
- Bugfix: `VOID` fields read & written incorrectly
- Bugfix: `.gui` files can now be opened

ERF editor
----------
- Columns can be sorted by clicking on the column header (sort by resource name, type or size)
- Description editor is only displayed for MOD an HAK files (not for ERFs)
- Type `sav` has been removed from the type selector (`.sav` and `.nwm` files have the erf type MOD)

TLK editor
----------
- Bugfix (Ornedan): position field accepts values in the user entry range


Release 11
==========
- Removed mnemonics for toolbar buttons with icons
- Menu shortcuts should work on Mac OS
- Improved readme.html ( marginally improved - wouldn't call it 'good' ;) )

TLK & 2DA editor
----------------
- Added search and replace dialog

TLK (NwnLanguage)
-----------------
- Added support for Chinese, Japanese and Korean tlk files

2DA editor
----------
- DEFAULT: field is now retained ( but there is no way to edit it the editor )

GFF (NwnLanguage)
-----------------
- Support for CExoLocStrings in Chinese, Japanese or Korean


Release 10a
===========
- Replaced icons with icons from the CrystalSVG icon set for KDE
- TlkEdit is now released under the GNU GPL

TLK editor
----------
- Added command line option for overriding character encoding.
  Character encoding used by tlkedit can be set with `-Dtlkedit.encoding=<encoding name>`
  command line option (hint: edit `tlkedit.bat`) e.g.
  ```console
  java  -Dtlkedit.encoding=cp1251 tlkedit.jar
  ```
  for a complete list of encodings supported by java 1.4.2 see
  http://java.sun.com/j2se/1.4.2/docs/guide/intl/encoding.doc.html


Release 10
==========
- TlkEdit now comes packaged as an executable jar file which will install the application.
  On Windows you should be able to run the jar file by double-clicking it or if double-clicking
  starts a zip utility, right-click the jar file and select "open with->javaw"
  in case that doesn't work either you need to launch it from the command line with:
  ```console
  java -jar TlkEditInstallerR10.jar
  ```
- Changed modifier key for most hotkeys to CTRL (so `CTRL+S` instead of `ALT+S` will save the file etc.),
  the `ALT+?` key bindings still remain though

TLK editor
----------
- Cells will be resized to display only one line after editing
- Cell editors now respond properly to ESC ( abort edit ) and F2 ( start edit ) (table lost focus when pressing ESC)

2DA editor
----------
- Behaviour of cell editors changed ( see tlk editor )

src/org/jl/nwn/NwnLanguage.java
package org.jl.nwn;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.EnumMap;
import java.util.EnumSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public enum NwnLanguage{    
    
    ENGLISH( "English", Version.NWN1, 0, Locale.ENGLISH ),
    FRENCH( "French", Version.NWN1, 1, Locale.FRENCH ),
    GERMAN( "German", Version.NWN1, 2, Locale.GERMAN ),
    ITALIAN( "Italian", Version.NWN1, 3, Locale.ITALIAN ),
    SPANISH( "Spanish", Version.NWN1, 4, new Locale("es") ),
    POLISH( "Polish", Version.NWN1, 5, new Locale("pl") ),
    
    KOREAN( "Korean", Version.NWN1, 128, Locale.KOREAN ),
    CHIN_TRAD( "Chinese, Traditional", Version.NWN1, 129, Locale.TRADITIONAL_CHINESE ),
    CHIN_SIMP( "Chinese, Simplified", Version.NWN1, 130, Locale.SIMPLIFIED_CHINESE ),
    JAPANESE( "Japanese", Version.NWN1, 131, Locale.JAPANESE ),
    
    NWN2_TOKENID( "GffToken", Version.NWN2, -1, Locale.ENGLISH ),    
    NWN2_ENGLISH( "English", Version.NWN2, 0, Locale.ENGLISH ),
    NWN2_FRENCH( "French", Version.NWN2, 1, Locale.FRENCH),
    NWN2_GERMAN( "German", Version.NWN2, 2, Locale.GERMAN ),
    NWN2_ITALIAN( "Italian", Version.NWN2, 3, Locale.ITALIAN ),
    NWN2_SPANISH( "Spanish", Version.NWN2, 4, new Locale("es") ),
    NWN2_POLISH( "Polish", Version.NWN2, 5, new Locale("pl") ),
    NWN2_6( "UnknownNwn2(RU?)", Version.NWN2, 6, new Locale("ru") ),    
    NWN2_KOREAN( "Korean", Version.NWN2, 128, Locale.KOREAN ),
    NWN2_CHIN_TRAD( "Chinese, Traditional", Version.NWN2, 129, Locale.TRADITIONAL_CHINESE ),
    NWN2_CHIN_SIMP( "Chinese, Simplified", Version.NWN2, 130, Locale.SIMPLIFIED_CHINESE ),
    NWN2_JAPANESE( "Japanese", Version.NWN2, 131, Locale.JAPANESE ),
    
    WI_DEBUG( "Debug", Version.WITCHER, 0, Locale.ENGLISH ),
    WI_ENGLISH( "English", Version.WITCHER, 1, Locale.ENGLISH),
    WI_FINALENGLISH("FinalEnglish", Version.WITCHER, 2, Locale.ENGLISH),
    WI_FINALENGLISH_SHORT("FinalEnglish_Short", Version.WITCHER, 3, Locale.ENGLISH),
    WI_POLISH("Polish", Version.WITCHER, 5, new Locale("pl")),
    WI_GERMAN("German", Version.WITCHER, 10, Locale.GERMAN),

    WI_FRENCH("French", Version.WITCHER, 11, Locale.FRENCH),
    WI_SPANISH("Spanish", Version.WITCHER, 12, new Locale("es")),
    WI_ITALIAN("Italian", Version.WITCHER, 13, Locale.ITALIAN),
    WI_RUSSIAN("Russian", Version.WITCHER, 14, new Locale("ru")),
    WI_CZECH("Czech", Version.WITCHER, 15, new Locale("cs")),
    WI_HUNGARIAN("Hungarian", Version.WITCHER, 16, new Locale("hu")),
    
    WI_KOREAN("Korean", Version.WITCHER, 20, Locale.KOREAN ),
    WI_CHIN_TRAD("Chinese, Traditional", Version.WITCHER, 21, Locale.TRADITIONAL_CHINESE),
    WI_CHIN_SIMP("Chinese, Simplified", Version.WITCHER, 22, Locale.SIMPLIFIED_CHINESE );

    private String name;
    private int code;
    private Version version;
    // language locale
    private Locale locale;
    
    private NwnLanguage( String name, Version v, int code, Locale locale ){
        this.name=name;
        this.code=code;
        this.version=v;
        this.locale=locale;
    }
    
    public String getName(){
        return name;
    }
    
    public int getCode(){
        return code;
    }
    
    public Version getVersion(){
        return version;
    }
    
    public Locale getLocale(){
        return locale;
    }
    
    public String getEncoding(){
        return encodings.get(this);
    }
    
    public String toString(){
        return name;
    }
    
    public static NwnLanguage find( Version v, int code ){
        Set<NwnLanguage> s = null;
        switch (v){
            case NWN1 : s=NWN1_LANGUAGES; break;
            case NWN2 : s=NWN2_LANGUAGES; break;
            case WITCHER : s=WITCHER_LANGUAGES; break;
        }
        for ( NwnLanguage l : s )
            if ( l.getCode() == code )
                return l;
        String errmsg = "Error : unsupported language, language code : "
                + code + ", Version : " + v;
        System.err.println(errmsg);
        throw new Error(
                errmsg);
    }
    
    public static List<NwnLanguage> findAll(Version v){
        List<NwnLanguage> l = new ArrayList<NwnLanguage>();
        for (NwnLanguage lang : EnumSet.allOf(NwnLanguage.class))
            if (lang.getVersion().equals(v))
                l.add(lang);
        return l;
    }
    
    public static void setEncoding(NwnLanguage l, String enc){
        encodings.put(l, enc);
    }
    
    private static EnumMap<NwnLanguage, String> encodings;
    
    public final static Set<NwnLanguage> NWN1_LANGUAGES = Collections.unmodifiableSet(
            EnumSet.range(NwnLanguage.ENGLISH, NwnLanguage.JAPANESE));
    
    public final static Set<NwnLanguage> NWN2_LANGUAGES = Collections.unmodifiableSet(
            EnumSet.range(NwnLanguage.NWN2_TOKENID, NwnLanguage.NWN2_JAPANESE));
    
    public final static Set<NwnLanguage> WITCHER_LANGUAGES = Collections.unmodifiableSet(
            EnumSet.range(NwnLanguage.WI_DEBUG, NwnLanguage.WI_CHIN_SIMP));
    
    public final static List<NwnLanguage> LANGUAGES = Arrays.asList(NWN1_LANGUAGES.toArray(new NwnLanguage[0]));
    
    static{
        encodings = new EnumMap<NwnLanguage, String>(NwnLanguage.class);
        encodings.put( ENGLISH, "windows-1252" );
        encodings.put( FRENCH, "windows-1252" );
        encodings.put( GERMAN, "windows-1252" );
        encodings.put( ITALIAN, "windows-1252" );
        encodings.put( SPANISH, "windows-1252" );
        encodings.put( POLISH, "windows-1250" );
        encodings.put( KOREAN, "MS949" );
        encodings.put( CHIN_TRAD, "MS950" );
        encodings.put( CHIN_SIMP, "MS936" );
        encodings.put( JAPANESE, "MS932" );
        
        for (NwnLanguage l : NWN2_LANGUAGES )
            encodings.put( l, "UTF-8" );
        /*
        encodings.put( WI_DEBUG, "windows-1252" );
        encodings.put( WI_ENGLISH, "windows-1252" );
        encodings.put( WI_FINALENGLISH, "windows-1252" );
        encodings.put( WI_FINALENGLISH_SHORT, "windows-1252" );
        encodings.put( WI_POLISH, "windows-1250" );
        encodings.put( WI_FRENCH, "windows-1252" );
        encodings.put( WI_GERMAN, "windows-1252" );
        encodings.put( WI_ITALIAN, "windows-1252" );
        encodings.put( WI_SPANISH, "windows-1252" );
        //encodings.put( WI_RUSSIAN, "windows-1250" );
        encodings.put( WI_RUSSIAN, "UTF-8" );
        encodings.put( WI_CZECH, "windows-1251" );
        encodings.put( WI_HUNGARIAN, "windows-1251" );
        //encodings.put( WI_KOREAN, "MS949" );
        encodings.put( WI_KOREAN, "UTF-8" );
        encodings.put( WI_CHINESETRAD, "UTF-8" );
        //encodings.put( WI_CHINESETRAD, "MS950" );
        //encodings.put( WI_CHINESESIMP, "MS936" );
        encodings.put( WI_CHINESESIMP, "UTF-8" );
        */
        // languages.2da lists different codepages, however the tlk files in the
        // game seem to use UTF-8
        for (NwnLanguage l : WITCHER_LANGUAGES )
            encodings.put( l, "UTF-8" );
        

        
        String enc = null;
        if ( ( enc = System.getProperty( "tlkedit.charsetOverride" )) != null ){
            String[] values = enc.split(";");
            try{
                for ( int p = 0; p < values.length; p++ ){
                    String[] triple = values[p].split(":");
                    Version v = Enum.valueOf(Version.class, triple[0]);
                    NwnLanguage l = find( v, Integer.parseInt(triple[1]) );
                    String encoding = triple[2];
                    encodings.put(l, encoding);
                    System.out.printf("nwnlanguage.java encoding for %s (%s) set to %s\n",
                            l.getName(), v, encoding);
                }
            } catch ( Throwable t ){
                System.err.println(
                        "invalid format for tlkedit.encodingOverride : " +
                        enc);
                System.err.println(t);
            }
        }
        
        for (NwnLanguage l : values() ){
            System.out.printf("NwnLanguage %s : %s[%s], %s, %s:%s\n", l.name(), l.version.name(), l.getCode(), l.getName(), l.getLocale(), l.getLocale().getDisplayLanguage());
        }
        
    }
}
